collab-document/src/document.rs

```diff
@@ -386,8 +386,8 @@ impl Document {
     K: Into<Origin>,
     F: Fn(HashMap<ClientID, DocumentAwarenessState>) + Send + Sync + 'static,
   {
-    self.collab.get_awareness().on_update_with(key, move |awareness, e, _| {
-      if let Ok(full_update) = awareness.update_with_clients(e.all_changes()) {
```

Comment:
This was the issue: `e.all_changes()` means all added/removed/updated clients **changed** as part of this observer call, not all known clients. Because of this if the client state didn't change, it was not emitted upstream and didn't appear in the view refresh.